architecture/2025-06-compute-extensibility-feature-spec.md

```diff
@@ -0,0 +1,469 @@
+# Topic: Compute Platform Extensibility for Radius
+
+* **Author**: Will Tsai (@willtsai)
+
+## Terms and definitions
+
+| Term | Definition |
+|------|------------|
+| **Compute platform** | An environment where applications can be deployed and run, such as Kubernetes, Azure Container Instances (ACI), etc. |
+| **Core types** | Built-in resource types provided by Radius, including `containers`, `gateways`, `secretStores`, `environments`, and `applications`. |
+| **User-defined type (UDT)** | A custom resource type defined separately from Radius core types and loaded into Radius without requiring a new Radius release. |
+| **Recipe** | A set of instructions that provisions a Radius resource to a Radius environment. Recipes are implemented in Bicep or Terraform. |
+| **Resource Provider (RP)** | A component responsible for handling create, read, update, delete, and list (CRUDL) operations for a specific resource type. |
+
+## Topic Summary
+<!-- A paragraph or two to summarize the topic area. Just define it in summary form so we all know what it is. -->
+Radius will be enhanced to support multiple compute platforms, secret stores, and gateway resources through a recipe-based extensibility model. This approach decouples Radius's core logic from platform-specific provisioning code. Core resource types (`containers`, `gateways`, and `secretStores`) will be implemented as User-Defined Types (UDTs) and will allow platform engineers to register Bicep or Terraform recipes for them. Radius will provide default recipes for Kubernetes and Azure Container Instances (ACI), but platform engineers can use, modify, or replace these to customize how Radius provisions resources to different environments, or to add support for entirely new platforms without requiring changes to Radius core.
+
+### Top level goals
+<!-- At the most basic level, what are we trying to accomplish? -->
+- Provide platform engineers with the ability to deploy Radius applications to specific platforms other than the default Kubernetes, such as Azure Container Instances (ACI), and other compute, secret store, and gateway platforms.
+- Provide abstraction punch-through mechanisms to allow platform engineers and developers to leverage platform-specific features and configurations not directly supported by Radius core (e.g. confidential containers).
+- Provide a recipe-based platform engineering experience that is consistent for user-defined types and core types.
+- Expand the ability to provision a single application definition to multiple clouds by adding the capability to provision to multiple compute platforms, secret stores, and gateway types.
+- Architecturally separate Radius core logic from platform-specific provisioning code.
+- Enable community-provided extensions to support new platforms without Radius code changes.
+
+### Non-goals (out of scope)
+<!-- What are we explicitly not trying to accomplish? -->
+- Running the Radius control plane on a non-Kubernetes platform.
```

Comment:
Does this mean that with extensibility and core types as UDTs we shouldn't expect to be able to host UCP in a different platform?